src/block/Hopper.php

```diff
@@ -96,8 +103,238 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
-		//TODO
+		$this->position->getWorld()->scheduleDelayedBlockUpdate($this->position, 1);
+
+		$tile = $this->position->getWorld()->getTile($this->position);
+		if(!$tile instanceof TileHopper){
+			return;
+		}
+
+		$transferCooldown = $tile->getTransferCooldown();
+		if($transferCooldown > 0){
+			$transferCooldown--;
+			$tile->setTransferCooldown($transferCooldown);
+		}
+
+		if($this->isPowered()){
+			// If a hopper is powered it is deactivated and won't push, pull or pick up items.
+			return;
+		}
+
+		if($transferCooldown > 0){
+			// Skipping this tick because the hopper is still on cooldown of 8 ticks.
+			return;
+		}
+
+		$success = $this->push($tile);
+		// Hoppers that have a container above them, won't try to pick up items.
+		$origin = $this->position->getWorld()->getTile($this->position->getSide(Facing::UP));
+		if($origin instanceof Container){
+			$success |= $this->pull($tile, $origin);
+		}else{
+			$success |= $this->pickup($tile);
+		}
+		// The cooldown is only set back to 8 ticks if the hopper has done anything.
+		if((bool) $success){
+			$tile->setTransferCooldown(8);
+		}
+	}
+
+	/**
+	 * This function handles pushing items from the hopper to a tile in the direction the hopper is facing.
+	 * Returns true if an item was successfully pushed or false on failure.
+	 */
+	private function push(TileHopper $tile) : bool{
+		if(count($tile->getInventory()->getContents()) === 0){
+			return false;
+		}
+		$destination = $this->position->getWorld()->getTile($this->position->getSide($this->facing));
+		if($destination === null){
+			return false;
+		}
+
+		for($slot = 0; $slot < $tile->getInventory()->getSize(); $slot++){
+			$item = $tile->getInventory()->getItem($slot);
+			if($item->isNull()){
+				continue;
+			}
+
+			// Hoppers interact differently when pushing into different kinds of tiles.
+			// TODO Composter
+			// TODO Brewing Stand
+			// TODO Jukebox (improve)
+			if($destination instanceof TileFurnace){
+				// If the hopper is facing down, it will push every item to the furnace's smelting slot, even items that aren't smeltable.
+				// If the hopper is facing in any other direction, it will only push items that can be used as fuel to the furnace's fuel slot.
+				if($this->facing === Facing::DOWN){
+					// ID of the smelting slot
+					$slotInFurnace = 0;
+					$itemInFurnace = $destination->getInventory()->getSmelting();
+				}else{
+					if($item->getFuelTime() === 0){
+						continue;
+					}
+					// ID of the fuel slot
+					$slotInFurnace = 1;
+					$itemInFurnace = $destination->getInventory()->getFuel();
+				}
+				if(!$itemInFurnace->isNull()){
+					if($itemInFurnace->getCount() >= $itemInFurnace->getMaxStackSize()){
+						return false;
+					}
+					if(!$itemInFurnace->canStackWith($item)){
+						continue;
+					}
+					$item->pop();
+					$itemInFurnace->setCount($itemInFurnace->getCount() + 1);
+				}else{
+					$itemInFurnace = $item->pop();
+				}
+
+				// TODO event on item inventory switch
```

Comment:
I saw this in your main PR TODOs, but I believe that events should be included before this PR is merged.

Reply:
Dylan already pointed out that this PR should only contain the logic itself and no events.
Also, the event for block picking up an item only exists since this commit https://github.com/pmmp/PocketMine-MP/commit/2139171a5525a89061cbadfa30f3f3f1243973b4 and we don't have an event for inventory interaction (InventoryTransactionEvent is not possible for this case, explained in https://github.com/pmmp/PocketMine-MP/issues/4405)